unified_planning/model/state.py

```diff
@@ -13,12 +13,108 @@
 # limitations under the License.
 #
 
+from typing import Dict, List, Optional
 import unified_planning as up
+from unified_planning.exceptions import UPUsageError, UPValueError
 
 
-class State:
-    '''This is an abstract class representing a classical state'''
+class ROState:
+    '''This is an abstract class representing a classical Read Only state'''
 
-    def get_value(self, f: 'up.model.FNode'):
-        '''Returns the value assignment for the given fluent expression'''
+    def get_value(self, value: 'up.model.FNode') -> 'up.model.FNode':
+        '''
+        This method retrieves the value in the state.
+        NOTE that the searched value must be set in the state.
+        :params value: The value searched for in the state.
+        :return: The set value.
+        '''
         raise NotImplementedError
+
+
+class RWState(ROState):
+    '''This is an abstract class representing a classical Read/Write state'''
+
+    def set_values(self, new_values: Dict['up.model.FNode', 'up.model.FNode']) -> 'RWState':
+        '''
+        Returns a different UPRWState in which every value in new_values.keys() is evaluated as his mapping
+        in new the new_values dict and every other value is evaluated as in self.
+        :param new_values: The dictionary that contains the values that need to be updated in the new State.
+        :return: The new State created.
+        '''
+        raise NotImplementedError
+
+
+class UPRWState(RWState):
+    '''
+    unified_planning implementation of the RWState interface.
+    This class has a field "MAX_ANCESTORS" set to 20.
+
+    The higher this numer is, the less memory the data structure will use.
+    The lower this number is, the less time the data structure will need to retrieve a value.
+
+    To set your own number just extend this class and re-define the MAX_ANCESTORS value. It must be > 0
+    '''
+    MAX_ANCESTORS: int = 20
+    '''This class is the unified_planning implementation of the RWState interface.'''
+    def __init__(self, values: Dict['up.model.FNode', 'up.model.FNode'], father: Optional['UPRWState'] = None):
+        if type(self).MAX_ANCESTORS < 1:
+            raise UPValueError(f'The max_ancestor field of a class extending UPRWState must be > 0: in the class {type(self)} it is set to {type(self).MAX_ANCESTORS}')
+        self._father = father
+        self._values = values
+        if father is None:
+            self._ancestors = 0
+        else:
+            self._ancestors = father._ancestors + 1
+
+    def __repr__(self) -> str:
+        current_instance = self
+        retval = ''
+        while current_instance is not None:
+            retval = f'{retval}\n{str(current_instance._values)}'
```

Comment:
This is very inefficient in python, because strings are immutable. Use a list and then concatenate it